### ISSUE TITLE
Feature Request: Allow a -RetryCount on an It, and run the test until it passes or the retry count is exhausted 

### ISSUE THREAD

**Issue by TravisEz13** [pester/Pester#1332]
Allow a -RetryCount on an It, and run the test until it passes or the retry count is exhausted

We have some tests that require internet connectivity.  
Writing custom retry logic for each test is annoying.  
It would be nice if this was just built into pester.  
No, checking for what the error was.  
If it is not safe to retry on all errors,
then the test should be refactored or rewritten.

**Comment by nohwnd** [comment]
Thanks for the idea, I see this would be useful for environment tests, but not for the general testing. I don't think this will be added any time soon, but I will put it to 5.x milestone so we can review it again. And will try to provide a workaround solution in the meantime. Is that okay?

**Comment by TravisEz13** [comment]
That’s fine.  

We have a workaround, but it requires… custom work in each test:
https://github.com/PowerShell/PowerShell/blob/a702f689749bd4886097b078ef6a3caa638d2752/test/tools/Modules/HelpersCommon/HelpersCommon.psm1#L3

**Comment by nohwnd** [comment]
[Your usage](https://github.com/PowerShell/PowerShell/blob/a702f689749bd4886097b078ef6a3caa638d2752/test/powershell/Modules/Microsoft.PowerShell.Core/Get-PSHostProcessInfo.Tests.ps1#L42) is a bit different from what I was thinking about. 

In your use case you are  waiting for something to become true so you can test it. In my head I was thinking about retrying the whole test case till it passes. Slight difference but might be important. 

If I did this I would model it around exceptions and would catch and swallow errors if there are more retries to do. This would mean that the whole body of the test would repeat, not just the precondition:

```powershell
It "Should list Windows PowerShell process" -Skip:(!$IsWindows) {
    Retry -Count 3 -Interval 1000 {
        $psProcess = Get-PSHostProcessInfo | Where-Object { $_.ProcessName -eq "powershell" }
        $psProcess.Count | Should -BeGreaterOrEqual 1
        $psProcess.ProcessId | Should -Contain $powershell.id
    }
}
```

To me this seems less error prone, and with reasonable defaults it is almost painless even without native support in Pester. This is also the only way the Pester could implement this for It (if you wanted your approach then the retry would have to be added to BeforeAll/BeforeEach).

Personally I would specify the retries as a number or repetitions and not a timeout, because then the number of retries does not depend on the amount of work that the test does. And also because timeout suggests that the test will be cancelled, if it takes longer than the specified time which is not true if the code blocks.

**Comment by TravisEz13** [comment]
@nohwnd Yes, the function was written for a specific case, and then found to have general usage.  So the parameter are probably not ideal.  I opened this issue, basically to try to generalize the pattern.

I was asking for the whole test case, but built in function would work as well, especially with reasonable defaults.  Honestly, a built in function, would probably handle more cases.
